packages/language/src/language-server/document-symbol-builder.ts

```diff
@@ -0,0 +1,326 @@
+/**
+ * This program and the accompanying materials are made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-v20.html
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Copyright Contributors to the Zowe Project.
+ *
+ */
+
+import {
+  DocumentSymbol,
+  SymbolKind,
+  Position,
+} from "vscode-languageserver-types";
+import { TextDocument } from "vscode-languageserver-textdocument";
+import { DeclaredItem, Statement, SyntaxKind } from "../syntax-tree/ast";
+import { Range, rangeToLSP, getSyntaxNodeRange, tokenToRange } from "./types";
+import { IToken } from "chevrotain";
+import { CstNodeKind } from "../syntax-tree/cst";
+interface LevelSymbol {
+  level: number | null;
+  symbol: DocumentSymbol;
+}
+
+interface SymbolBuilder {
+  canHandle(token: IToken): boolean;
+  buildSymbols(
+    token: IToken,
+    elementTokens: IToken[],
+    textDocument: TextDocument,
+    childSymbols: DocumentSymbol[],
+  ): DocumentSymbol[];
+}
+
+class ProcedureSymbolBuilder implements SymbolBuilder {
+  canHandle(token: IToken): boolean {
+    return (
+      token.tokenTypeIdx == CstNodeKind.EntryStatement_Variable &&
+      token.payload?.element.kind === SyntaxKind.ProcedureStatement
+    );
+  }
+
+  buildSymbols(
+    token: IToken,
+    elementTokens: IToken[],
+    textDocument: TextDocument,
+    childSymbols: DocumentSymbol[],
+  ): DocumentSymbol[] {
+    const labelPrefixStatement = token.payload?.element.container as Statement;
+    const procedureName =
+      labelPrefixStatement.labels.map((label) => label.name).join(" ") ??
+      "<unknown>";
+    const range = getSyntaxNodeRange(labelPrefixStatement.labels[0]) ?? {
+      start: 0,
+      end: 0,
+    };
```

Comment:
Suggestion: We shouldn't fall back to `0:0` as the range for the syntax node. Can we use the `token.offset` position instead?